{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MoNHr8Atk1s2"
      },
      "source": [
        "# Final Project of Introduction to Bioinformatics\n",
        "\n",
        "## Comparative Mitochondrial Genomics and Phylogenetic Analysis\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "9rDhGHcSk1s3"
      },
      "source": [
        "This task focuses on the comparative analysis of mitochondrial genomes from different species, primarily birds, mammals, and insects. The aim is to understand the evolutionary relationships between these species by analyzing and comparing their mitochondrial DNA, which is about 16,000 base pairs in length. You will use advanced computational methods to construct phylogenetic trees and delve into the ecological and anthropological insights that can be gleaned from this data. This project is designed to provide a comprehensive understanding of mitochondrial genomics, its importance in evolutionary biology, and its applications in broader scientific contexts.\n",
        "\n",
        "You will learn:\n",
        "\n",
        "- Techniques for aligning and comparing mitochondrial DNA sequences.\n",
        "- How to construct and interpret phylogenetic trees using advanced computational methods.\n",
        "- The application of mitochondrial genomics in understanding ecological interactions and human evolutionary history.\n",
        "\n",
        "#### Task Roadmap\n",
        "\n",
        "1. **Mitochondrial Genome Comparison**:\n",
        "   - Align mitochondrial DNA sequences from the provided dataset.\n",
        "   - Analyze these sequences to identify similarities and differences across species.\n",
        "\n",
        "2. **Phylogenetic Analysis Using Advanced Methods**:\n",
        "   - Apply Maximum Likelihood (ML) and Bayesian Inference methods, utilizing tools like `ETE Toolkit`, `DendroPy`, `BEAST`, or `PyRate`.\n",
        "   - Compare the trees generated by these methods to understand how different approaches can lead to different interpretations of the data.\n",
        "\n",
        "3. **Cross-Disciplinary Applications (Bonus)**:\n",
        "   - **Ecology**: Examine how mitochondrial DNA analysis can reveal information about species adaptation, migration, and conservation. This involves understanding how genetic variation within and between species can inform ecological strategies and conservation efforts.\n",
        "   - **Anthropology**: Investigate the use of mitochondrial DNA in tracing human evolution and migration patterns. This includes studying the mitochondrial DNA of mammals in your dataset to draw parallels with human evolutionary studies.\n",
        "\n",
        "### Data Sources\n",
        "\n",
        "The mitochondrial DNA data for birds, mammals, and insects will be provided to you. This dataset has been curated to facilitate a comprehensive comparative analysis and is essential for the completion of your phylogenetic studies.\n",
        "\n",
        "### Useful Resources and Material\n",
        "\n",
        "- [Mitochondrial DNA - Wikipedia](https://en.wikipedia.org/wiki/Mitochondrial_DNA): A general introduction to the structure, function, origin, and diversity of mitochondrial DNA, as well as its applications in various fields such as medicine, forensics, and anthropology.\n",
        "- [Mitochondrial DNA Analysis: Introduction, Methods, and Applications](https://bioinfo.cd-genomics.com/mitochondrial-dna-analysis-introduction-methods-and-applications.html): An explanation of the basics of mitochondrial DNA sequencing, bioinformatics analysis, heteroplasmy, and advantages of mitochondrial DNA analysis over nuclear DNA analysis.\n",
        "- [Phylogenetic Tree- Definition, Types, Steps, Methods, Uses - Microbe Notes](https://microbenotes.com/phylogenetic-tree/): A coverage of the concepts and methods of phylogenetic tree construction, including the types of phylogenetic trees, the steps involved in phylogenetic analysis, the main methods of phylogenetic inference, and the applications of phylogenetic trees in various disciplines.\n",
        "- [Phylogenetics - Wikipedia](https://en.wikipedia.org/wiki/Phylogenetics): An overview of the field of phylogenetics, which is the study of the evolutionary history and relationships among or within groups of organisms. It also discusses the data sources, models, algorithms, software, and challenges of phylogenetic analysis.\n",
        "- [ETE Toolkit](http://etetoolkit.org/): A Python library for manipulating, analyzing, and visualizing phylogenetic trees. It supports various formats, methods, and tools for phylogenetic analysis, such as alignment, tree inference, tree comparison, tree annotation, and tree visualization.\n",
        "- [DendroPy](https://dendropy.org/): Another Python library for phylogenetic computing. It provides a comprehensive API for working with phylogenetic data structures, such as trees, characters, and networks. It also offers a rich set of functions for simulation, manipulation, analysis, and annotation of phylogenetic data."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zsCLKWGFk1s4"
      },
      "source": [
        "### Exploration and Reflection\n",
        "\n",
        "As we proceed with our analysis of mitochondrial DNA for phylogenetic tree construction, it is valuable to contemplate a few questions. These inquiries aim to facilitate a more thorough understanding of the roles and characteristics of mitochondrial DNA in the context of evolutionary biology:\n",
        "\n",
        "1. **Maternal Inheritance and Its Implications**: How does the maternal inheritance of mitochondrial DNA simplify our understanding of evolutionary lineage compared to nuclear DNA, which undergoes recombination? What unique insights can this aspect provide in tracing the evolutionary history of species?\n",
        "\n",
        "2. **Mutation Rate and Evolutionary Insights**: Mitochondrial DNA mutates at a faster rate than nuclear DNA. How does this characteristic make mtDNA a more sensitive tool for detecting recent evolutionary events and relationships among closely related species? Can you think of any specific scenarios or studies where this property of mtDNA has been particularly instrumental?\n",
        "\n",
        "Reflect on these questions as you work through the project, and consider how the properties of mitochondrial DNA enhance its value and applicability in evolutionary biology and beyond. Provide your answer either in this notebook, or in your report (if you had one).\n",
        "\n",
        "<blockquote style=\"font-family:Arial; color:red; font-size:16px; border-left:0px solid red; padding: 10px;\">\n",
        "    <strong>Don't forget to answer these questions!</strong>\n",
        "</blockquote>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Me64tFmyk1s4"
      },
      "source": [
        "### Step 0: Installing Necessary Packages"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 2,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ocwTpgjIk1s4",
        "outputId": "c1a41d51-7304-4916-8af3-381cdd14635a"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "biopython (1.81) is installed\n",
            "pandas (2.1.2) is installed\n",
            "numpy (1.24.3) is installed\n"
          ]
        }
      ],
      "source": [
        "import sys\n",
        "import subprocess\n",
        "import pkg_resources\n",
        "\n",
        "def install(package):\n",
        "    subprocess.check_call([sys.executable, \"-m\", \"pip\", \"install\", package])\n",
        "\n",
        "REQUIRED_PACKAGES = [\n",
        "    'biopython',\n",
        "    'pandas',\n",
        "    'numpy'\n",
        "]\n",
        "\n",
        "for package in REQUIRED_PACKAGES:\n",
        "    try:\n",
        "        dist = pkg_resources.get_distribution(package)\n",
        "        print('{} ({}) is installed'.format(dist.key, dist.version))\n",
        "    except pkg_resources.DistributionNotFound:\n",
        "        print('{} is NOT installed'.format(package))\n",
        "        install(package)\n",
        "        print('{} was successfully installed.'.format(package))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 29,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xmsUlIcuk1s5",
        "outputId": "5e2dd7d3-9115-47e6-c490-086611bb5bc4"
      },
      "outputs": [
        {
          "data": {
            "text/plain": [
              "True"
            ]
          },
          "execution_count": 29,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Import necessary libraries.\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "from dotenv import load_dotenv\n",
        "\n",
        "# Load environment variables from .env file\n",
        "load_dotenv(dotenv_path='../../.env')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "PiUM9k_Kk1s5"
      },
      "source": [
        "### Step 1: Dataset Expansion\n",
        "\n",
        "Our first task is to augment our dataset with additional species. This involves engaging with the NCBI database to retrieve mitochondrial DNA sequences.\n",
        "\n",
        "#### Instructions:\n",
        "\n",
        "- **Species Selection**: Identify and choose 10 additional species to include in your dataset. Aim for a diverse selection to enrich your phylogenetic analysis.\n",
        "\n",
        "- **Querying NCBI Database**: Use the NCBI database to locate mitochondrial DNA sequences for your chosen species. While you can manually search on the [NCBI website](https://www.ncbi.nlm.nih.gov/), consider automating this process through their API for a more efficient approach.\n",
        "    - **Example Query**: As a starting point, you might use a query like `\"mitochondrion[Filter] AND (your_species_name[Organism])` to find specific mtDNA sequences. Adjust the query parameters according to your species selection.\n",
        "    - **Documentation**: Familiarize yourself with the [NCBI API documentation](https://www.ncbi.nlm.nih.gov/books/NBK25497/) for detailed guidance on constructing queries.\n",
        "\n",
        "- **Using NCBI Website**: You are welcome to use the NCBI website for this task. If you do so, document each step of your process clearly in your task report. This should include the species names, search terms used, and how you determined the relevant sequences to include.\n",
        "\n",
        "- **Bonus Opportunity**: Implementing an automated, methodological approach using the NCBI API and relevant Python packages to add all 10 records in your dataset will earn you a 50% bonus for this section. Your method should be structured and replicable, demonstrating a systematic approach to data collection.\n",
        "\n",
        "Remember, the goal is to methodically expand your dataset with relevant mtDNA sequences, paving the way for insightful phylogenetic analysis."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        },
        "id": "o1jzAmQBk1s6",
        "outputId": "ada500ff-0c2c-4b3f-cc71-6cb3a7fc4a9b"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>taxo_id</th>\n",
              "      <th>specie</th>\n",
              "      <th>blast_name</th>\n",
              "      <th>genbank_common_name</th>\n",
              "      <th>accession_number</th>\n",
              "      <th>mtDNA</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>8945</td>\n",
              "      <td>Eudynamys scolopaceus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Asian koel</td>\n",
              "      <td>NC_060520</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>7460</td>\n",
              "      <td>Apis mellifera</td>\n",
              "      <td>Bees</td>\n",
              "      <td>honey bee</td>\n",
              "      <td>NC_051932</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>36300</td>\n",
              "      <td>Pelecanus crispus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Dalmatian pelican</td>\n",
              "      <td>OR620163</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>10116</td>\n",
              "      <td>Rattus norvegicus</td>\n",
              "      <td>Rodents</td>\n",
              "      <td>Norway rat</td>\n",
              "      <td>NC_001665</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_001665</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>9031</td>\n",
              "      <td>Gallus gallus</td>\n",
              "      <td>birds</td>\n",
              "      <td>Gallus gallus</td>\n",
              "      <td>NC_053523</td>\n",
              "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "   taxo_id                 specie blast_name genbank_common_name  \\\n",
              "0     8945  Eudynamys scolopaceus      birds          Asian koel   \n",
              "1     7460         Apis mellifera       Bees           honey bee   \n",
              "2    36300      Pelecanus crispus      birds   Dalmatian pelican   \n",
              "3    10116      Rattus norvegicus    Rodents          Norway rat   \n",
              "4     9031          Gallus gallus      birds       Gallus gallus   \n",
              "\n",
              "  accession_number                                              mtDNA  \n",
              "0        NC_060520  https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...  \n",
              "1        NC_051932  https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...  \n",
              "2         OR620163    https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1  \n",
              "3        NC_001665     https://www.ncbi.nlm.nih.gov/nuccore/NC_001665  \n",
              "4        NC_053523   https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1  "
            ]
          },
          "execution_count": 4,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "dataset = pd.read_csv('./dataset/species.csv')\n",
        "\n",
        "dataset.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6JvfOKwYk1s6",
        "outputId": "af539cae-6ab7-4cf3-9586-b8111ff629e1"
      },
      "outputs": [],
      "source": [
        "from Bio import Entrez\n",
        "namelist = [\"Lagopus muta\",\n",
        "            \"Pteropus alecto\",\n",
        "            \"Himaloaesalus gaoligongshanus\",\n",
        "            \"Lima vulgaris\",\n",
        "            \"Gobio gobio\",\n",
        "            \"Ennucula tenuis\",\n",
        "            \"Mactra cumingii\",\n",
        "            \"Raeta pulchella\",\n",
        "            \"Cardita variegata\",\n",
        "            \"Capreolus capreolus\"]\n",
        "Entrez.email = \"serohrt@gmail.com\"\n",
        "for name in namelist:\n",
        "    handle = Entrez.esearch(db=\"taxonomy\", term=name)\n",
        "    records = Entrez.read(handle)\n",
        "    taxo_id = records[\"IdList\"][0]\n",
        "    handle_fetch = Entrez.efetch(db=\"taxonomy\", id=taxo_id,retmode = \"xml\")\n",
        "    recordsT = Entrez.read(handle_fetch)\n",
        "    specie = recordsT[0][\"ScientificName\"]\n",
        "    other_names = recordsT[0].get(\"OtherNames\")\n",
        "    if \"GenbankCommonName\" in other_names:\n",
        "        genbank_common_name = other_names[\"GenbankCommonName\"]\n",
        "    else:\n",
        "        genbank_common_name = \"NULL\"\n",
        "    genbank_handle = Entrez.esearch(db=\"nucleotide\", term = name + \" mitochondrion complete genome\")\n",
        "    uid = Entrez.read(genbank_handle)[\"IdList\"][0]\n",
        "    nuc_handle = Entrez.efetch(db=\"nucleotide\", id=uid, retmode=\"xml\", rettype=\"gb\")\n",
        "    accession_number = Entrez.read(nuc_handle)[0]['GBSeq_locus']\n",
        "    mtDNA = f\"https://www.ncbi.nlm.nih.gov/nucleotide/{accession_number}\"\n",
        "    dataset = dataset.append({\"taxo_id\": taxo_id, \n",
        "                              \"specie\": specie,\n",
        "                              \"blast_name\": specie,\n",
        "                              \"genbank_common_name\": genbank_common_name,\n",
        "                              \"accession_number\": accession_number,\n",
        "                              \"mtDNA\": mtDNA}, ignore_index=True)\n",
        "dataset.to_csv('expanded_dataset.csv')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "VO68BnQck1s6"
      },
      "source": [
        "#### Checking Data Consistency\n",
        "\n",
        "The code block is designed to check the accuracy of a biological dataset. It examines each entry in a CSV file, focusing on taxonomy IDs, species names, GenBank accession numbers, and mitochondrial DNA links. It uses the NCBI's Entrez system to ensure taxonomy IDs correspond to the correct species and confirms the mitochondrial DNA links are accurate. The script also checks GenBank accession numbers against the provided links. This method is useful for maintaining the accuracy of current data and **might help in adding new entries to the database.**"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {
        "id": "_R_mjLqtk1s6"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import requests\n",
        "from Bio import Entrez\n",
        "import os\n",
        "from io import BytesIO\n",
        "from requests.adapters import HTTPAdapter\n",
        "from urllib3.util.retry import Retry\n",
        "\n",
        "# Set your email and API key for NCBI\n",
        "entrez_email = os.getenv('ENTREZ_EMAIL')\n",
        "entrez_key = os.getenv('ENTREZ_API_KEY')\n",
        "Entrez.email = entrez_email\n",
        "\n",
        "def fetch_entrez_record(db, id, rettype, retmode):\n",
        "    \"\"\"Fetch record from NCBI Entrez with retries.\"\"\"\n",
        "    url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db={db}&id={id}&rettype={rettype}&retmode={retmode}\"\n",
        "    session = requests.Session()\n",
        "    retries = Retry(total=5, backoff_factor=1, status_forcelist=[502, 503, 504])\n",
        "    session.mount('https://', HTTPAdapter(max_retries=retries))\n",
        "\n",
        "    try:\n",
        "        response = session.get(url)\n",
        "        response.raise_for_status()\n",
        "        if db == \"nucleotide\":\n",
        "            return response.text\n",
        "        return response.content\n",
        "    except requests.exceptions.HTTPError as err:\n",
        "        print(f\"HTTP error: {err}\")\n",
        "    except requests.exceptions.ConnectionError as err:\n",
        "        print(f\"Connection error: {err}\")\n",
        "    return None\n",
        "\n",
        "def verify_taxonomy_id(taxo_id, species_name):\n",
        "    \"\"\"Verify taxonomy ID against species name.\"\"\"\n",
        "    xml_data = fetch_entrez_record(\"taxonomy\", taxo_id, \"xml\", \"xml\")\n",
        "    if xml_data:\n",
        "        # Convert bytes data to a binary file-like object\n",
        "        xml_data_io = BytesIO(xml_data)\n",
        "        records = Entrez.read(xml_data_io)\n",
        "        return records[0]['ScientificName'].lower() == species_name.lower()\n",
        "    return False\n",
        "\n",
        "def verify_mitochondrial_dna(accession_number):\n",
        "    gb_data = fetch_entrez_record(\"nucleotide\", accession_number, \"gb\", \"text\")\n",
        "    return \"mitochondrion\" in gb_data.lower() if gb_data else False\n",
        "\n",
        "def extract_accession_from_link(link):\n",
        "    return link.split('/')[-1].split('.')[0]\n",
        "\n",
        "def check_dataset_consistency(file_path):\n",
        "    species_df = pd.read_csv(file_path)\n",
        "\n",
        "    for index, row in species_df.iterrows():\n",
        "        taxonomy_id = str(row['taxo_id'])\n",
        "        species_name = row['specie']\n",
        "        accession_number = row['accession_number']\n",
        "        mtDNA_link = row['mtDNA']\n",
        "        extracted_accession = extract_accession_from_link(mtDNA_link)\n",
        "\n",
        "        taxonomy_check = verify_taxonomy_id(taxonomy_id, species_name)\n",
        "        accession_match = (accession_number == extracted_accession)\n",
        "        mitochondrial_check = verify_mitochondrial_dna(accession_number)\n",
        "\n",
        "        print(f\"Row {index}: Taxonomy Check: {taxonomy_check}, Accession Match: {accession_match}, Mitochondrial Check: {mitochondrial_check}\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "sNrLUTfhk1s6",
        "outputId": "ffe89690-9524-4e3d-c38f-b5b95788bc53"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Row 0: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 1: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 2: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 3: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 4: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 5: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 6: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 7: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 8: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 9: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 10: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 11: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 12: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 13: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 14: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 15: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 16: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 17: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 18: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 19: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 20: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 21: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
            "Row 22: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n"
          ]
        }
      ],
      "source": [
        "check_dataset_consistency('./dataset/species.csv')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "iaF6rrn2k1s6"
      },
      "source": [
        "### Step 3: Sequence Download and Preparation\n",
        "\n",
        "The next step in our project involves downloading the mitochondrial DNA sequences for each species and preparing them for analysis.\n",
        "\n",
        "#### Instructions:\n",
        "\n",
        "- **Download mtDNA Sequences**: Write a script to download the mtDNA sequences from the links provided in your dataset. The sequences should be in FASTA format, which is the standard for nucleotide sequences.\n",
        "\n",
        "- **Sequence Labeling**: Properly label each sequence within the FASTA file. This header, starting with '>', should include the species name and any other relevant information (e.g., `>Eudynamys_scolopaceus_NC_060520`). This is crucial for identifying the sequences in subsequent analysis.\n",
        "\n",
        "- **Concatenate Sequences**:\n",
        "    - Create a script to concatenate all downloaded sequences into a single `.fasta` or `.fna` file.\n",
        "    - Ensure each sequence in the file is clearly separated by its header line, which is important for differentiating the sequences of various species.\n",
        "\n",
        "#### Tips for Writing the Download and Concatenation Script:\n",
        "- Use Python libraries such as `httpx` or `requests`, or any other tool you prefer for downloading sequences. For processing FASTA files you can use a wide range of tools. One recommended option is `Biopython` library.\n",
        "- Use a loop to go through each link in the dataset, download the sequence, and append it to your concatenated file.\n",
        "- Maintain the format integrity of the FASTA file, ensuring each sequence is correctly associated with its header.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 12,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lzvZQXfik1s7",
        "outputId": "a9f22404-ccde-44f3-a8de-b9e5716ef0f5"
      },
      "outputs": [],
      "source": [
        "import requests\n",
        "import pandas as pd\n",
        "from Bio import SeqIO\n",
        "\n",
        "# Create a label for the sequence\n",
        "def create_label(id, specie):\n",
        "    id = id[1:]\n",
        "    return f\">{specie[0]}_{specie[1]}_{id}\\n\"\n",
        "\n",
        "# Create the proper URL out of the dataset URL to download the mtDNA sequence\n",
        "def create_url(url):\n",
        "    accession_num = url.split('/')[-1].split('.')[0]\n",
        "    db = url.split('/')[-2]\n",
        "    return f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db={db}&id={accession_num}&rettype=fasta&retmode=text\"\n",
        "\n",
        "# Download the mtDNA sequence from the URL\n",
        "def download_mtDNA(url):\n",
        "    link = create_url(url)\n",
        "    response = requests.get(link)\n",
        "\n",
        "    if response.status_code == 200:\n",
        "        return response.text[:-1]\n",
        "    else:\n",
        "        return None\n",
        "\n",
        "# Save the labeled sequences to a fasta file\n",
        "def save_fasta_file(sequences):    \n",
        "    with open(\"./outputs/sequences.fasta\", 'w') as f:\n",
        "        for entry in sequences:\n",
        "            f.write(entry)\n",
        "\n",
        "# Iterate through the dataset and download the mtDNA sequences\n",
        "sequences = []\n",
        "for index, row in dataset.iterrows():\n",
        "    fasta_entry = download_mtDNA(row['mtDNA'])\n",
        "    if fasta_entry:\n",
        "        # Parse the fasta entry and create a label out of it\n",
        "        parsed = fasta_entry.split(' ')\n",
        "        # Remove \"UNVERIFIED\" strings from the parsed list\n",
        "        for p in parsed:\n",
        "            if p.startswith(\"UNVERIFIED\"):\n",
        "                parsed.remove(p)\n",
        "        label = create_label(parsed[0], parsed[1:3])\n",
        "        # Remove \"genome\" string and the newline characters from the genome sequence\n",
        "        genome = \"\".join(parsed[-1].split('\\n')[1:]) + \"\\n\"\n",
        "        # Add them to the sequences list\n",
        "        sequences += [label, genome]\n",
        "save_fasta_file(sequences)"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lp-ZM-0gk1s7"
      },
      "source": [
        "### Step 4: Sequence Alignment\n",
        "\n",
        "After downloading the mitochondrial DNA sequences, the next critical step is their alignment. This process allows us to compare the sequences and discern the evolutionary relationships among the species.\n",
        "\n",
        "#### Instructions:\n",
        "\n",
        "- **Select an Alignment Tool**: Choose one of the following alignment tools based on your project needs. Each tool has its strengths and is widely used in bioinformatics for multiple sequence alignment.\n",
        "\n",
        "1. **MAFFT**:\n",
        "    - **Brief Introduction**: MAFFT (Multiple Alignment using Fast Fourier Transform) is renowned for its speed and efficiency, particularly suitable for large datasets.\n",
        "    - **Resources**:\n",
        "        - [MAFFT Official Documentation](https://mafft.cbrc.jp/alignment/software/)\n",
        "        - [Example Usage on GitHub](https://github.com/MountainMan12/SARS-Cov2-phylo)\n",
        "        - [Relevant Notebook](https://colab.research.google.com/github/pb3lab/ibm3202/blob/master/tutorials/lab03_phylo.ipynb)\n",
        "\n",
        "2. **Clustal Omega**:\n",
        "    - **Brief Introduction**: Clustal Omega offers high-quality alignments and is user-friendly, ideal for those new to sequence alignment.\n",
        "    - **Resources**:\n",
        "        - [A Python wrapper around Clustal Omega](https://github.com/benchling/clustalo-python)\n",
        "        - [Clustal Omega Official Website](http://www.clustal.org/omega/)\n",
        "\n",
        "3. **MUSCLE**:\n",
        "    - **Brief Introduction**: MUSCLE (Multiple Sequence Comparison by Log-Expectation) is known for its balance between speed and accuracy, making it a versatile choice for various datasets.\n",
        "    - **Resources**:\n",
        "        - [MUSCLE Documentation](https://drive5.com/muscle5/manual/)\n",
        "\n",
        "- **Perform Sequence Alignment**: Utilize your chosen tool to align the downloaded mtDNA sequences. This alignment is foundational for the accurate construction of phylogenetic trees.\n",
        "\n",
        "- **Save Aligned Sequences**: After alignment, save the output in an appropriate format for further analysis in the subsequent steps of this project."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 13,
      "metadata": {
        "id": "F1BoE6-Ek1s7"
      },
      "outputs": [],
      "source": [
        "from Bio.Align.Applications import MafftCommandline\n",
        "\n",
        "def perform_alignment(input_file, output_file):\n",
        "    mafft_cline = MafftCommandline(input=input_file)\n",
        "    stdout, _ = mafft_cline()\n",
        "    with open(output_file, \"w\") as f:\n",
        "        f.write(stdout)\n",
        "\n",
        "perform_alignment(\"./outputs/sequences.fasta\", \"./outputs/aligned_sequences.fasta\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "34ZJVzSck1s7"
      },
      "source": [
        "### Step 5: Phylogenetic Tree Construction\n",
        "\n",
        "The next phase in our project involves constructing phylogenetic trees to visualize and analyze the evolutionary relationships among the species. We will use three distinct methods, each providing unique insights.\n",
        "\n",
        "#### Phylogenetic Tree Construction Methods:\n",
        "\n",
        "1. **Bayesian Inference Trees**:\n",
        "    - **Overview**: This method uses Bayesian statistics to estimate the likelihood of different evolutionary histories. It's particularly useful for its ability to estimate branch lengths and support values.\n",
        "    - **Tools**: MrBayes, BEAST\n",
        "        - MrBayes ([Official Website](https://nbisweden.github.io/MrBayes/manual.html/)) is widely recognized for its robustness in Bayesian inference.\n",
        "        - BEAST2 ([BEAST Software](https://www.beast2.org/)) is another powerful tool, offering advanced features for complex evolutionary models.\n",
        "\n",
        "2. **Maximum Likelihood Trees**:\n",
        "    - **Overview**: Maximum Likelihood methods evaluate tree topologies based on the likelihood of observed data given a tree model. It's known for its statistical rigor and accuracy.\n",
        "    - **Tools**: RAxML, PhyML\n",
        "        - RAxML ([RAxML GitHub](https://github.com/stamatak/standard-RAxML)) is preferred for large datasets due to its efficiency.\n",
        "        - PhyML ([PhyML Documentation](http://www.atgc-montpellier.fr/phyml/)) offers a balance of speed and accuracy, with a user-friendly interface.\n",
        "\n",
        "3. **Neighbor-Joining Trees**:\n",
        "    - **Overview**: The Neighbor-Joining method is a distance-based approach that constructs phylogenetic trees by evaluating the genetic distance between sequences. It is known for its speed and simplicity, making it well-suited for initial exploratory analyses.\n",
        "    - **Tools**:\n",
        "        - MEGA: A versatile tool specifically used here for constructing Neighbor-Joining trees. It's recognized for its ease of use and effectiveness in phylogenetic analysis. [MEGA Software](https://www.megasoftware.net/)\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 79,
      "metadata": {
        "id": "kqt_PuF9k1s7"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\n",
            "RAxML-NG v. 1.2.1 released on 22.12.2023 by The Exelixis Lab.\n",
            "Developed by: Alexey M. Kozlov and Alexandros Stamatakis.\n",
            "Contributors: Diego Darriba, Tomas Flouri, Benoit Morel, Sarah Lutteropp, Ben Bettisworth, Julia Haag, Anastasis Togkousidis.\n",
            "Latest version: https://github.com/amkozlov/raxml-ng\n",
            "Questions/problems/suggestions? Please visit: https://groups.google.com/forum/#!forum/raxml\n",
            "\n",
            "System: Apple M2 Pro, 12 cores, 16 GB RAM\n",
            "\n",
            "RAxML-NG was called at 07-Feb-2024 18:33:11 as follows:\n",
            "\n",
            "../../applications/raxml-ng --search1 --msa ./outputs/aligned_sequences.fasta --model GTR+G --prefix ./trees/ML/phylo_ML\n",
            "\n",
            "Analysis options:\n",
            "  run mode: ML tree search\n",
            "  start tree(s): parsimony (1)\n",
            "  random seed: 1707318191\n",
            "  tip-inner: OFF\n",
            "  pattern compression: ON\n",
            "  per-rate scalers: OFF\n",
            "  site repeats: ON\n",
            "  logLH epsilon: general: 10.000000, brlen-triplet: 1000.000000\n",
            "  fast spr radius: AUTO\n",
            "  spr subtree cutoff: 1.000000\n",
            "  fast CLV updates: ON\n",
            "  branch lengths: proportional (ML estimate, algorithm: NR-FAST)\n",
            "  SIMD kernels: SSE3\n",
            "  parallelization: coarse-grained (auto), PTHREADS (auto)\n",
            "\n",
            "WARNING: The model you specified on the command line (GTR+G) will be ignored \n",
            "         since the binary MSA file already contains a model definition.\n",
            "         If you want to change the model, please re-run RAxML-NG \n",
            "         with the original PHYLIP/FASTA alignment and --redo option.\n",
            "\n",
            "[00:00:00] Loading binary alignment from file: ./trees/ML/phylo_ML.raxml.rba\n",
            "[00:00:00] Alignment comprises 23 taxa, 1 partitions and 12539 patterns\n",
            "\n",
            "Partition 0: noname\n",
            "Model: GTR+FO+G4m\n",
            "Alignment sites / patterns: 35509 / 12539\n",
            "Gaps: 52.74 %\n",
            "Invariant sites: 50.17 %\n",
            "\n",
            "\n",
            "Parallelization scheme autoconfig: 1 worker(s) x 12 thread(s)\n",
            "\n",
            "Parallel reduction/worker buffer size: 1 KB  / 0 KB\n",
            "\n",
            "[00:00:00] NOTE: Resuming execution from checkpoint (logLH: -251296.71, ML trees: 0, bootstraps: 0)\n",
            "[00:00:00] Data distribution: max. partitions/sites/weight per thread: 1 / 1045 / 16720\n",
            "[00:00:00] Data distribution: max. searches per worker: 1\n",
            "\n",
            "Starting ML tree search with 1 distinct starting trees\n",
            "\n",
            "[00:00:00 -251296.706742] SPR radius for FAST iterations: 10 (autodetect)\n",
            "[00:00:00 -251296.706742] SLOW spr round 2 (radius: 10)\n",
            "[00:00:02 -251296.705816] SLOW spr round 3 (radius: 15)\n",
            "[00:00:02 -251296.705727] SLOW spr round 4 (radius: 20)\n",
            "[00:00:03 -251296.705714] Model parameter optimization (eps = 0.100000)\n",
            "\n",
            "[00:00:03] ML tree search #1, logLikelihood: -251296.461719\n",
            "\n",
            "\n",
            "Optimized model parameters:\n",
            "\n",
            "   Partition 0: noname\n",
            "   Rate heterogeneity: GAMMA (4 cats, mean),  alpha: 0.519087 (ML),  weights&rates: (0.250000,0.036978) (0.250000,0.264194) (0.250000,0.833674) (0.250000,2.865154) \n",
            "   Base frequencies (ML): 0.386664 0.233385 0.094347 0.285605 \n",
            "   Substitution rates (ML): 1.659870 5.188869 0.896990 0.794498 5.281863 1.000000 \n",
            "\n",
            "\n",
            "Final LogLikelihood: -251296.461719\n",
            "\n",
            "AIC score: 502696.923439 / AICc score: 502697.078899 / BIC score: 503137.755595\n",
            "Free parameters (model + branch lengths): 52\n",
            "\n",
            "Best ML tree saved to: /Users/nikanvasei/Desktop/University/Semester_5/Bio/Project/Introduction-To-Bioinformatics/tasks/mitochondrial-comparison-phylogenetic-construction/trees/ML/phylo_ML.raxml.bestTree\n",
            "Optimized model saved to: /Users/nikanvasei/Desktop/University/Semester_5/Bio/Project/Introduction-To-Bioinformatics/tasks/mitochondrial-comparison-phylogenetic-construction/trees/ML/phylo_ML.raxml.bestModel\n",
            "\n",
            "Execution log saved to: /Users/nikanvasei/Desktop/University/Semester_5/Bio/Project/Introduction-To-Bioinformatics/tasks/mitochondrial-comparison-phylogenetic-construction/trees/ML/phylo_ML.raxml.log\n",
            "\n",
            "Analysis started: 07-Feb-2024 18:33:11 / finished: 07-Feb-2024 18:33:15\n",
            "\n",
            "Elapsed time: 3.553 seconds (this run) / 15.437 seconds (total with restarts)\n",
            "\n",
            "MrBayes output: \n",
            "\n",
            "                            MrBayes 3.2.7a arm\n",
            "\n",
            "                      (Bayesian Analysis of Phylogeny)\n",
            "\n",
            "                             (Parallel version)\n",
            "                         (1 processors available)\n",
            "\n",
            "              Distributed under the GNU General Public License\n",
            "\n",
            "\n",
            "               Type \"help\" or \"help <command>\" for information\n",
            "                     on the commands that are available.\n",
            "\n",
            "                   Type \"about\" for authorship and general\n",
            "                       information about the program.\n",
            "\n",
            "\n",
            "MrBayes > \n",
            "   Executing file \"./trees/BI/mb_bi.nex\"\n",
            "   UNIX line termination\n",
            "   Longest line length = 35550\n",
            "   Parsing file\n",
            "   Expecting NEXUS formatted file\n",
            "   Reading data block\n",
            "      Allocated taxon set\n",
            "      Allocated matrix\n",
            "      Defining new matrix with 23 taxa and 35510 characters\n",
            "      Data is Dna\n",
            "      Gaps coded as -\n",
            "      Missing data coded as ?\n",
            "      Taxon  1 -> Eudynamys_scolopaceus_NC_060520_1\n",
            "      Taxon  2 -> Apis_mellifera_NC_051932_1\n",
            "      Taxon  3 -> Pelecanus_crispus_OR620163_1\n",
            "      Taxon  4 -> Rattus_norvegicus_NC_001665_2\n",
            "      Taxon  5 -> Gallus_gallus_NC_053523_1\n",
            "      Taxon  6 -> Myotis_lucifugus_NC_029849_1\n",
            "      Taxon  7 -> Tribolium_castaneum_NC_003081_2\n",
            "      Taxon  8 -> Anser_cygnoides_NC_023832_1\n",
            "      Taxon  9 -> Ovis_aries_NC_001941_1\n",
            "      Taxon 10 -> Tyto_alba_MZ318036_1\n",
            "      Taxon 11 -> Drosophila_melanogaster_NC_024511_2\n",
            "      Taxon 12 -> Apteryx_australis_MN356385_1\n",
            "      Taxon 13 -> Spheniscus_demersus_NC_022817_1\n",
            "      Taxon 14 -> Corvus_brachyrhynchos_NC_026461_1\n",
            "      Taxon 15 -> Mus_musculus_NC_005089_1\n",
            "      Taxon 16 -> Dromaius_novaehollandiae_NC_002784_1\n",
            "      Taxon 17 -> Anas_platyrhynchos_NC_009684_1\n",
            "      Taxon 18 -> Pteropus_vampyrus_NC_026542_1\n",
            "      Taxon 19 -> Struthio_camelus_NC_002785_1\n",
            "      Taxon 20 -> Aquila_nipalensis_NC_045042_1\n",
            "      Taxon 21 -> Cygnus_cygnus_NC_027095_1\n",
            "      Taxon 22 -> Taeniopygia_guttata_NC_007897_1\n",
            "      Taxon 23 -> Aptenodytes_forsteri_NC_027938_1\n",
            "      Successfully read matrix\n",
            "      Setting default partition (does not divide up characters)\n",
            "      Setting model defaults\n",
            "      Seed (for generating default start values) = 1707318195\n",
            "      Setting output file names to \"./trees/BI/mb_bi.nex.run<i>.<p|t>\"\n",
            "   Exiting data block\n",
            "   Reading mrbayes block\n",
            "      Setting autoclose to yes\n",
            "      Setting nowarnings to yes\n",
            "      Setting Nst to 6\n",
            "      Setting Rates to Invgamma\n",
            "      Successfully set likelihood model parameters\n",
            "      Setting Statefreqpr to Dirichlet(1.00,1.00,1.00,1.00)\n",
            "      Successfully set prior model parameters\n",
            "      Setting Ratepr to Variable [Dirichlet(..,1,..)] for partition 1\n",
            "      Successfully set prior model parameters\n",
            "      Setting number of generations to 25000\n",
            "      Using relative burnin (a fraction of samples discarded).\n",
            "      Setting burnin fraction to 0.25\n",
            "      Setting print frequency to 1000\n",
            "      Setting sample frequency to 1000\n",
            "      Setting number of chains to 4\n",
            "      Successfully set chain parameters\n",
            "      Running Markov chain\n",
            "      MCMC stamp = 1083172049\n",
            "      Seed = 1488961653\n",
            "      Swapseed = 1707318195\n",
            "      Model settings:\n",
            "\n",
            "         Data not partitioned --\n",
            "            Datatype  = DNA\n",
            "            Nucmodel  = 4by4\n",
            "            Nst       = 6\n",
            "                        Substitution rates, expressed as proportions\n",
            "                        of the rate sum, have a Dirichlet prior\n",
            "                        (1.00,1.00,1.00,1.00,1.00,1.00)\n",
            "            Covarion  = No\n",
            "            # States  = 4\n",
            "                        State frequencies have a Dirichlet prior\n",
            "                        (1.00,1.00,1.00,1.00)\n",
            "            Rates     = Invgamma\n",
            "                        The distribution is approximated using 4 categories.\n",
            "                        Shape parameter is exponentially\n",
            "                        distributed with parameter (1.00).\n",
            "                        Proportion of invariable sites is uniformly dist-\n",
            "                        ributed on the interval (0.00,1.00).\n",
            "\n",
            "      Active parameters: \n",
            "\n",
            "         Parameters\n",
            "         ---------------------\n",
            "         Revmat              1\n",
            "         Statefreq           2\n",
            "         Shape               3\n",
            "         Pinvar              4\n",
            "         Ratemultiplier      5\n",
            "         Topology            6\n",
            "         Brlens              7\n",
            "         ---------------------\n",
            "\n",
            "         1 --  Parameter  = Revmat\n",
            "               Type       = Rates of reversible rate matrix\n",
            "               Prior      = Dirichlet(1.00,1.00,1.00,1.00,1.00,1.00)\n",
            "\n",
            "         2 --  Parameter  = Pi\n",
            "               Type       = Stationary state frequencies\n",
            "               Prior      = Dirichlet\n",
            "\n",
            "         3 --  Parameter  = Alpha\n",
            "               Type       = Shape of scaled gamma distribution of site rates\n",
            "               Prior      = Exponential(1.00)\n",
            "\n",
            "         4 --  Parameter  = Pinvar\n",
            "               Type       = Proportion of invariable sites\n",
            "               Prior      = Uniform(0.00,1.00)\n",
            "\n",
            "         5 --  Parameter  = Ratemultiplier\n",
            "               Type       = Partition-specific rate multiplier\n",
            "               Prior      = Dirichlet(1.00)\n",
            "\n",
            "         6 --  Parameter  = Tau\n",
            "               Type       = Topology\n",
            "               Prior      = All topologies equally probable a priori\n",
            "               Subparam.  = V\n",
            "\n",
            "         7 --  Parameter  = V\n",
            "               Type       = Branch lengths\n",
            "               Prior      = Unconstrained:GammaDir(1.0,0.1000,1.0,1.0)\n",
            "\n",
            "\n",
            "      Number of chains per MPI processor = 8\n",
            "\n",
            "      The MCMC sampler will use the following moves:\n",
            "         With prob.  Chain will use move\n",
            "            0.90 %   Dirichlet(Revmat)\n",
            "            0.90 %   Slider(Revmat)\n",
            "            0.90 %   Dirichlet(Pi)\n",
            "            0.90 %   Slider(Pi)\n",
            "            1.80 %   Multiplier(Alpha)\n",
            "            1.80 %   Slider(Pinvar)\n",
            "            1.35 %   Dirichlet(Ratemultiplier)\n",
            "            1.35 %   Slider(Ratemultiplier)\n",
            "            9.01 %   ExtSPR(Tau,V)\n",
            "            9.01 %   ExtTBR(Tau,V)\n",
            "            9.01 %   NNI(Tau,V)\n",
            "            9.01 %   ParsSPR(Tau,V)\n",
            "           36.04 %   Multiplier(V)\n",
            "           12.61 %   Nodeslider(V)\n",
            "            5.41 %   TLMultiplier(V)\n",
            "\n",
            "      Division 1 has 12541 unique site patterns\n",
            "      Initializing conditional likelihoods\n",
            "      Using standard non-SSE likelihood calculator for division 1 (single-precision)\n",
            "      Initializing invariable-site conditional likelihoods\n",
            "\n",
            "      Initial log likelihoods and log prior probs for run 1:\n",
            "         Chain 1 -- -378758.200837 -- 69.461457\n",
            "         Chain 2 -- -356418.303161 -- 69.461457\n",
            "         Chain 3 -- -386002.635277 -- 69.461457\n",
            "         Chain 4 -- -382763.702525 -- 69.461457\n",
            "\n",
            "      Initial log likelihoods and log prior probs for run 2:\n",
            "         Chain 1 -- -376796.611253 -- 69.461457\n",
            "         Chain 2 -- -373055.937630 -- 69.461457\n",
            "         Chain 3 -- -380126.502612 -- 69.461457\n",
            "         Chain 4 -- -377322.649173 -- 69.461457\n",
            "\n",
            "\n",
            "      Using a relative burnin of 25.0 % for diagnostics\n",
            "\n",
            "      Chain results (25000 generations requested):\n",
            "\n",
            "         0 -- [-378758.201] (-356418.303) (-386002.635) (-382763.703) * [-376796.611] (-373055.938) (-380126.503) (-377322.649) (...0 remote chains...) \n",
            "      1000 -- [-257784.957] (-259285.083) (-258543.180) (-260613.753) * [-256813.852] (-262637.267) (-262885.367) (-261596.322) (...0 remote chains...) -- 0:10:48\n",
            "      2000 -- [-252915.534] (-253619.266) (-254186.021) (-254907.582) * [-252617.643] (-257794.815) (-255124.101) (-254334.373) (...0 remote chains...) -- 0:09:58\n",
            "      3000 -- (-251797.258) [-251813.115] (-253386.917) (-253774.121) * [-252240.483] (-255504.216) (-253544.771) (-252628.017) (...0 remote chains...) -- 0:09:32\n",
            "      4000 -- [-251574.289] (-251596.312) (-252495.627) (-252145.911) * (-252004.482) (-253866.520) (-252566.899) [-251756.725] (...0 remote chains...) -- 0:09:00\n",
            "      5000 -- (-251413.009) [-251408.494] (-251944.069) (-251547.081) * (-251650.243) (-252881.968) (-251590.185) [-251448.089] (...0 remote chains...) -- 0:08:32\n",
            "\n",
            "      Average standard deviation of split frequencies: 0.147850\n",
            "\n",
            "      6000 -- (-251383.371) [-251331.327] (-251720.771) (-251436.153) * (-251444.789) (-251997.499) (-251501.688) [-251417.309] (...0 remote chains...) -- 0:08:07\n",
            "      7000 -- (-251379.521) [-251319.077] (-251532.085) (-251384.008) * (-251426.643) (-251500.971) [-251431.833] (-251402.928) (...0 remote chains...) -- 0:07:40\n",
            "      8000 -- (-251376.297) [-251326.736] (-251412.436) (-251362.648) * (-251414.317) (-251443.390) (-251435.027) [-251375.884] (...0 remote chains...) -- 0:07:13\n",
            "      9000 -- (-251372.735) [-251320.254] (-251399.790) (-251344.123) * (-251406.390) (-251391.140) (-251391.744) [-251365.074] (...0 remote chains...) -- 0:06:48\n",
            "      10000 -- (-251367.061) [-251313.614] (-251394.329) (-251339.855) * (-251353.821) [-251361.899] (-251376.156) (-251376.816) (...0 remote chains...) -- 0:06:22\n",
            "\n",
            "      Average standard deviation of split frequencies: 0.180943\n",
            "\n",
            "      11000 -- (-251364.725) [-251317.806] (-251384.499) (-251343.456) * [-251342.298] (-251357.825) (-251364.135) (-251374.575) (...0 remote chains...) -- 0:05:57\n",
            "      12000 -- (-251351.860) [-251316.953] (-251379.425) (-251347.926) * (-251356.789) [-251332.793] (-251357.854) (-251384.148) (...0 remote chains...) -- 0:05:32\n",
            "      13000 -- (-251360.650) [-251310.195] (-251373.203) (-251354.040) * (-251329.665) [-251329.392] (-251351.587) (-251387.175) (...0 remote chains...) -- 0:05:07\n",
            "      14000 -- (-251356.518) [-251313.342] (-251373.539) (-251354.094) * (-251323.804) [-251323.491] (-251339.242) (-251384.693) (...0 remote chains...) -- 0:04:42\n",
            "      15000 -- (-251357.095) [-251315.538] (-251364.727) (-251347.340) * [-251312.810] (-251327.131) (-251342.719) (-251373.339) (...0 remote chains...) -- 0:04:16\n",
            "\n",
            "      Average standard deviation of split frequencies: 0.145048\n",
            "\n",
            "      16000 -- (-251340.060) [-251309.070] (-251379.571) (-251347.943) * (-251311.373) [-251325.083] (-251342.925) (-251373.617) (...0 remote chains...) -- 0:03:50\n",
            "      17000 -- (-251336.245) [-251314.431] (-251360.668) (-251333.550) * (-251313.603) [-251324.207] (-251351.489) (-251364.613) (...0 remote chains...) -- 0:03:25\n",
            "      18000 -- (-251348.027) (-251316.263) (-251366.680) [-251311.422] * (-251317.124) [-251314.700] (-251345.994) (-251364.009) (...0 remote chains...) -- 0:02:59\n",
            "      19000 -- (-251353.200) (-251321.657) (-251361.305) [-251316.247] * (-251318.916) [-251318.859] (-251344.476) (-251359.177) (...0 remote chains...) -- 0:02:33\n",
            "      20000 -- (-251358.491) (-251308.285) (-251374.130) [-251326.354] * (-251318.725) [-251308.741] (-251343.596) (-251352.484) (...0 remote chains...) -- 0:02:08\n",
            "\n",
            "      Average standard deviation of split frequencies: 0.095187\n",
            "\n",
            "      21000 -- (-251355.503) [-251310.125] (-251367.680) (-251327.618) * (-251318.481) [-251319.815] (-251357.805) (-251357.653) (...0 remote chains...) -- 0:01:42\n",
            "      22000 -- (-251360.305) (-251324.537) (-251362.318) [-251317.329] * (-251320.773) [-251319.724] (-251348.148) (-251358.370) (...0 remote chains...) -- 0:01:16\n",
            "      23000 -- (-251346.605) (-251311.189) (-251362.238) [-251319.216] * [-251308.292] (-251328.095) (-251354.556) (-251360.189) (...0 remote chains...) -- 0:00:51\n",
            "      24000 -- (-251345.032) [-251309.413] (-251360.108) (-251307.549) * (-251316.782) [-251320.513] (-251350.017) (-251364.351) (...0 remote chains...) -- 0:00:25\n",
            "      25000 -- (-251346.910) [-251312.519] (-251363.558) (-251318.628) * (-251314.390) [-251318.515] (-251335.736) (-251358.765) (...0 remote chains...) -- 0:00:00\n",
            "\n",
            "      Average standard deviation of split frequencies: 0.059832\n",
            "\n",
            "      Analysis completed in 10 mins 42 seconds\n",
            "      Analysis used 634.72 seconds of CPU time on processor 0\n",
            "      Likelihood of best state for \"cold\" chain of run 1 was -251304.28\n",
            "      Likelihood of best state for \"cold\" chain of run 2 was -251305.64\n",
            "\n",
            "      Acceptance rates for the moves in the \"cold\" chain of run 1:\n",
            "         With prob.   (last 100)   chain accepted proposals by move\n",
            "             9.6 %     ( 17 %)     Dirichlet(Revmat)\n",
            "            18.7 %     (  7 %)     Slider(Revmat)\n",
            "             2.9 %     (  0 %)     Dirichlet(Pi)\n",
            "             7.7 %     (  1 %)     Slider(Pi)\n",
            "             6.6 %     (  4 %)     Multiplier(Alpha)\n",
            "            15.7 %     ( 12 %)     Slider(Pinvar)\n",
            "            40.8 %     ( 69 %)     Dirichlet(Ratemultiplier)\n",
            "            34.5 %     ( 31 %)     Slider(Ratemultiplier)\n",
            "             0.6 %     (  0 %)     ExtSPR(Tau,V)\n",
            "             0.7 %     (  0 %)     ExtTBR(Tau,V)\n",
            "             1.0 %     (  0 %)     NNI(Tau,V)\n",
            "             0.5 %     (  0 %)     ParsSPR(Tau,V)\n",
            "            20.9 %     ( 17 %)     Multiplier(V)\n",
            "             7.4 %     (  4 %)     Nodeslider(V)\n",
            "             2.8 %     (  3 %)     TLMultiplier(V)\n",
            "\n",
            "      Acceptance rates for the moves in the \"cold\" chain of run 2:\n",
            "         With prob.   (last 100)   chain accepted proposals by move\n",
            "             5.7 %     (  0 %)     Dirichlet(Revmat)\n",
            "            24.0 %     ( 16 %)     Slider(Revmat)\n",
            "             3.6 %     (  0 %)     Dirichlet(Pi)\n",
            "             9.8 %     (  6 %)     Slider(Pi)\n",
            "             7.5 %     (  6 %)     Multiplier(Alpha)\n",
            "            14.7 %     ( 10 %)     Slider(Pinvar)\n",
            "            10.7 %     (  9 %)     Dirichlet(Ratemultiplier)\n",
            "            37.2 %     ( 39 %)     Slider(Ratemultiplier)\n",
            "             0.7 %     (  0 %)     ExtSPR(Tau,V)\n",
            "             1.1 %     (  0 %)     ExtTBR(Tau,V)\n",
            "             0.7 %     (  0 %)     NNI(Tau,V)\n",
            "             0.9 %     (  0 %)     ParsSPR(Tau,V)\n",
            "            20.6 %     ( 19 %)     Multiplier(V)\n",
            "             6.3 %     (  6 %)     Nodeslider(V)\n",
            "             3.2 %     (  4 %)     TLMultiplier(V)\n",
            "\n",
            "      Chain swap information for run 1:\n",
            "\n",
            "                 1     2     3     4 \n",
            "           --------------------------\n",
            "         1 |        0.27  0.03  0.00 \n",
            "         2 |  4045        0.31  0.06 \n",
            "         3 |  4196  4209        0.27 \n",
            "         4 |  4158  4212  4180       \n",
            "\n",
            "      Chain swap information for run 2:\n",
            "\n",
            "                 1     2     3     4 \n",
            "           --------------------------\n",
            "         1 |        0.31  0.05  0.01 \n",
            "         2 |  4169        0.27  0.05 \n",
            "         3 |  4174  4111        0.36 \n",
            "         4 |  4146  4175  4225       \n",
            "\n",
            "      Upper diagonal: Proportion of successful state exchanges between chains\n",
            "      Lower diagonal: Number of attempted state exchanges between chains\n",
            "\n",
            "      Chain information:\n",
            "\n",
            "        ID -- Heat \n",
            "       -----------\n",
            "         1 -- 1.00  (cold chain)\n",
            "         2 -- 0.91 \n",
            "         3 -- 0.83 \n",
            "         4 -- 0.77 \n",
            "\n",
            "      Heat = 1 / (1 + T * (ID - 1))\n",
            "         (where T = 0.10 is the temperature and ID is the chain number)\n",
            "\n",
            "      Summarizing trees in files \"./trees/BI/mb_bi.nex.run1.t\" and \"./trees/BI/mb_bi.nex.run2.t\"\n",
            "      Using relative burnin ('relburnin=yes'), discarding the first 25 % of sampled trees\n",
            "      Writing statistics to files ./trees/BI/mb_bi.nex.<parts|tstat|vstat|trprobs|con>\n",
            "      Examining first file ...\n",
            "      Found one tree block in file \"./trees/BI/mb_bi.nex.run1.t\" with 26 trees in last block\n",
            "      Expecting the same number of trees in the last tree block of all files\n",
            "\n",
            "      Tree reading status:\n",
            "\n",
            "      0      10      20      30      40      50      60      70      80      90     100\n",
            "      v-------v-------v-------v-------v-------v-------v-------v-------v-------v-------v\n",
            "      *********************************************************************************\n",
            "\n",
            "      Read a total of 52 trees in 2 files (sampling 40 of them)\n",
            "         (Each file contained 26 trees of which 20 were sampled)\n",
            "                                                                                   \n",
            "      General explanation:                                                          \n",
            "                                                                                   \n",
            "      In an unrooted tree, a taxon bipartition (split) is specified by removing a   \n",
            "      branch, thereby dividing the species into those to the left and those to the  \n",
            "      right of the branch. Here, taxa to one side of the removed branch are denoted \n",
            "      '.' and those to the other side are denoted '*'. Specifically, the '.' symbol \n",
            "      is used for the taxa on the same side as the outgroup.                        \n",
            "                                                                                   \n",
            "      In a rooted or clock tree, the tree is rooted using the model and not by      \n",
            "      reference to an outgroup. Each bipartition therefore corresponds to a clade,  \n",
            "      that is, a group that includes all the descendants of a particular branch in  \n",
            "      the tree.  Taxa that are included in each clade are denoted using '*', and    \n",
            "      taxa that are not included are denoted using the '.' symbol.                  \n",
            "                                                                                   \n",
            "      The output first includes a key to all the bipartitions with frequency larger \n",
            "      or equual to (Minpartfreq) in at least one run. Minpartfreq is a parameter to \n",
            "      sumt command and currently it is set to 0.10.  This is followed by a table  \n",
            "      with statistics for the informative bipartitions (those including at least    \n",
            "      two taxa), sorted from highest to lowest probability. For each bipartition,   \n",
            "      the table gives the number of times the partition or split was observed in all\n",
            "      runs (#obs) and the posterior probability of the bipartition (Probab.), which \n",
            "      is the same as the split frequency. If several runs are summarized, this is   \n",
            "      followed by the minimum split frequency (Min(s)), the maximum frequency       \n",
            "      (Max(s)), and the standard deviation of frequencies (Stddev(s)) across runs.  \n",
            "      The latter value should approach 0 for all bipartitions as MCMC runs converge.\n",
            "                                                                                   \n",
            "      This is followed by a table summarizing branch lengths, node heights (if a    \n",
            "      clock model was used) and relaxed clock parameters (if a relaxed clock model  \n",
            "      was used). The mean, variance, and 95 % credible interval are given for each \n",
            "      of these parameters. If several runs are summarized, the potential scale      \n",
            "      reduction factor (PSRF) is also given; it should approach 1 as runs converge. \n",
            "      Node heights will take calibration points into account, if such points were   \n",
            "      used in the analysis.                                                         \n",
            "                                                                                    \n",
            "      Note that Stddev may be unreliable if the partition is not present in all     \n",
            "      runs (the last column indicates the number of runs that sampled the partition \n",
            "      if more than one run is summarized). The PSRF is not calculated at all if     \n",
            "      the partition is not present in all runs.The PSRF is also sensitive to small  \n",
            "      sample sizes and it should only be considered a rough guide to convergence    \n",
            "      since some of the assumptions allowing one to interpret it as a true potential\n",
            "      scale reduction factor are violated in MrBayes.                               \n",
            "                                                                                    \n",
            "      List of taxa in bipartitions:                                                 \n",
            "                                                                                   \n",
            "         1 -- Eudynamys_scolopaceus_NC_060520_1\n",
            "         2 -- Apis_mellifera_NC_051932_1\n",
            "         3 -- Pelecanus_crispus_OR620163_1\n",
            "         4 -- Rattus_norvegicus_NC_001665_2\n",
            "         5 -- Gallus_gallus_NC_053523_1\n",
            "         6 -- Myotis_lucifugus_NC_029849_1\n",
            "         7 -- Tribolium_castaneum_NC_003081_2\n",
            "         8 -- Anser_cygnoides_NC_023832_1\n",
            "         9 -- Ovis_aries_NC_001941_1\n",
            "        10 -- Tyto_alba_MZ318036_1\n",
            "        11 -- Drosophila_melanogaster_NC_024511_2\n",
            "        12 -- Apteryx_australis_MN356385_1\n",
            "        13 -- Spheniscus_demersus_NC_022817_1\n",
            "        14 -- Corvus_brachyrhynchos_NC_026461_1\n",
            "        15 -- Mus_musculus_NC_005089_1\n",
            "        16 -- Dromaius_novaehollandiae_NC_002784_1\n",
            "        17 -- Anas_platyrhynchos_NC_009684_1\n",
            "        18 -- Pteropus_vampyrus_NC_026542_1\n",
            "        19 -- Struthio_camelus_NC_002785_1\n",
            "        20 -- Aquila_nipalensis_NC_045042_1\n",
            "        21 -- Cygnus_cygnus_NC_027095_1\n",
            "        22 -- Taeniopygia_guttata_NC_007897_1\n",
            "        23 -- Aptenodytes_forsteri_NC_027938_1\n",
            "\n",
            "      Key to taxon bipartitions (saved to file \"./trees/BI/mb_bi.nex.parts\"):\n",
            "\n",
            "      ID -- Partition\n",
            "      -----------------------------\n",
            "       1 -- .**********************\n",
            "       2 -- .*.....................\n",
            "       3 -- ..*....................\n",
            "       4 -- ...*...................\n",
            "       5 -- ....*..................\n",
            "       6 -- .....*.................\n",
            "       7 -- ......*................\n",
            "       8 -- .......*...............\n",
            "       9 -- ........*..............\n",
            "      10 -- .........*.............\n",
            "      11 -- ..........*............\n",
            "      12 -- ...........*...........\n",
            "      13 -- ............*..........\n",
            "      14 -- .............*.........\n",
            "      15 -- ..............*........\n",
            "      16 -- ...............*.......\n",
            "      17 -- ................*......\n",
            "      18 -- .................*.....\n",
            "      19 -- ..................*....\n",
            "      20 -- ...................*...\n",
            "      21 -- ....................*..\n",
            "      22 -- .....................*.\n",
            "      23 -- ......................*\n",
            "      24 -- .....*..*........*.....\n",
            "      25 -- .*........*............\n",
            "      26 -- ...........*...*..*....\n",
            "      27 -- ...*..........*........\n",
            "      28 -- .*.*.**.*.*...*..*.....\n",
            "      29 -- ............*.........*\n",
            "      30 -- .*....*...*............\n",
            "      31 -- .....*...........*.....\n",
            "      32 -- .......*............*..\n",
            "      33 -- ....*..*........*...*..\n",
            "      34 -- .......*........*...*..\n",
            "      35 -- ...........*...*.......\n",
            "      36 -- ...*.*..*.....*..*.....\n",
            "      37 -- .............*.......*.\n",
            "      38 -- .*.*.**.*.*..**..*...*.\n",
            "      39 -- ....*..*...*...**.*.*..\n",
            "      40 -- ............*......*..*\n",
            "      41 -- .*.*.**.***..**..*...*.\n",
            "      42 -- .*.*********.******.**.\n",
            "      43 -- .*.********************\n",
            "      44 -- ..*.........*.........*\n",
            "      45 -- .........*.........*...\n",
            "      46 -- ..*......*..*......*..*\n",
            "      47 -- .************.*******.*\n",
            "      48 -- .*.*.**.*.**..**.**....\n",
            "      49 -- .*.******.**..*****.*..\n",
            "      -----------------------------\n",
            "\n",
            "      Summary statistics for informative taxon bipartitions\n",
            "         (saved to file \"./trees/BI/mb_bi.nex.tstat\"):\n",
            "\n",
            "      ID   #obs    Probab.     Sd(s)+      Min(s)      Max(s)   Nruns \n",
            "      ----------------------------------------------------------------\n",
            "      24    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      25    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      26    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      27    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      28    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      29    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      30    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      31    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      32    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      33    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      34    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      35    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      36    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      37    40    1.000000    0.000000    1.000000    1.000000    2\n",
            "      38    37    0.925000    0.106066    0.850000    1.000000    2\n",
            "      39    37    0.925000    0.106066    0.850000    1.000000    2\n",
            "      40    36    0.900000    0.141421    0.800000    1.000000    2\n",
            "      41    36    0.900000    0.141421    0.800000    1.000000    2\n",
            "      42    36    0.900000    0.141421    0.800000    1.000000    2\n",
            "      43    35    0.875000    0.176777    0.750000    1.000000    2\n",
            "      44     4    0.100000    0.141421    0.000000    0.200000    1 *\n",
            "      45     4    0.100000    0.141421    0.000000    0.200000    1 *\n",
            "      46     4    0.100000    0.141421    0.000000    0.200000    1 *\n",
            "      47     3    0.075000    0.106066    0.000000    0.150000    1 *\n",
            "      48     3    0.075000    0.106066    0.000000    0.150000    1 *\n",
            "      49     3    0.075000    0.106066    0.000000    0.150000    1 *\n",
            "      ----------------------------------------------------------------\n",
            "      + Convergence diagnostic (standard deviation of split frequencies)\n",
            "        should approach 0.0 as runs converge.\n",
            "\n",
            "      * The partition was not found in all runs so the values are unreliable\n",
            "\n",
            "      Summary statistics for branch and node parameters\n",
            "         (saved to file \"./trees/BI/mb_bi.nex.vstat\"):\n",
            "\n",
            "                                              95% HPD Interval\n",
            "                                            --------------------\n",
            "      Parameter      Mean       Variance     Lower       Upper       Median     PSRF+  Nruns\n",
            "      --------------------------------------------------------------------------------------\n",
            "      length[1]     0.192697    0.000926    0.142471    0.229174    0.199352    1.704    2\n",
            "      length[2]     0.986618    0.021914    0.787482    1.180519    0.995465    1.840    2\n",
            "      length[3]     0.115407    0.000364    0.074116    0.138551    0.121441    1.381    2\n",
            "      length[4]     0.117631    0.000309    0.087797    0.143303    0.122297    1.789    2\n",
            "      length[5]     0.171078    0.000655    0.128390    0.201096    0.177875    1.873    2\n",
            "      length[6]     0.228231    0.001025    0.174007    0.270525    0.241826    1.857    2\n",
            "      length[7]     0.405537    0.004475    0.303041    0.502649    0.419325    1.920    2\n",
            "      length[8]     0.048245    0.000054    0.035739    0.056763    0.051512    1.990    2\n",
            "      length[9]     0.176102    0.000658    0.125851    0.205493    0.184948    1.938    2\n",
            "      length[10]    0.220691    0.001126    0.156764    0.259528    0.231335    1.924    2\n",
            "      length[11]    0.170901    0.000667    0.127779    0.206119    0.174778    1.603    2\n",
            "      length[12]    0.118213    0.000332    0.084968    0.138283    0.124466    1.896    2\n",
            "      length[13]    0.064535    0.000081    0.050139    0.075936    0.067571    1.646    2\n",
            "      length[14]    0.119608    0.000359    0.089144    0.141878    0.125764    1.853    2\n",
            "      length[15]    0.150439    0.000531    0.112465    0.184204    0.156986    1.693    2\n",
            "      length[16]    0.100955    0.000209    0.076563    0.120334    0.106145    1.773    2\n",
            "      length[17]    0.072170    0.000122    0.050962    0.086157    0.072357    1.603    2\n",
            "      length[18]    0.169138    0.000610    0.125393    0.198877    0.174085    1.742    2\n",
            "      length[19]    0.116481    0.000277    0.089085    0.140040    0.119635    1.628    2\n",
            "      length[20]    0.138895    0.000514    0.102926    0.172289    0.143325    1.934    2\n",
            "      length[21]    0.048761    0.000053    0.036460    0.056452    0.048874    1.618    2\n",
            "      length[22]    0.105650    0.000283    0.074185    0.125035    0.109690    1.703    2\n",
            "      length[23]    0.065059    0.000090    0.048486    0.076112    0.068703    1.903    2\n",
            "      length[24]    0.103019    0.000253    0.074275    0.125137    0.104073    1.568    2\n",
            "      length[25]    0.250756    0.002024    0.154096    0.325192    0.254725    1.302    2\n",
            "      length[26]    0.079546    0.000221    0.049984    0.099562    0.085637    1.918    2\n",
            "      length[27]    0.182447    0.000789    0.136154    0.219007    0.187211    1.740    2\n",
            "      length[28]    0.293624    0.002072    0.197395    0.360555    0.303362    1.715    2\n",
            "      length[29]    0.045775    0.000051    0.033843    0.055675    0.046630    1.691    2\n",
            "      length[30]    2.138912    0.095194    1.621314    2.511746    2.262116    1.967    2\n",
            "      length[31]    0.051551    0.000091    0.034818    0.066733    0.052234    1.475    2\n",
            "      length[32]    0.026538    0.000019    0.018861    0.033141    0.026852    1.817    2\n",
            "      length[33]    0.031420    0.000034    0.020944    0.041384    0.031743    1.506    2\n",
            "      length[34]    0.084784    0.000142    0.063030    0.101010    0.087599    1.942    2\n",
            "      length[35]    0.028498    0.000021    0.019922    0.035010    0.028514    1.524    2\n",
            "      length[36]    0.274231    0.001988    0.193667    0.331278    0.292291    1.510    2\n",
            "      length[37]    0.073914    0.000244    0.050753    0.098392    0.071768    1.080    2\n",
            "      length[38]    0.033088    0.000040    0.023764    0.041541    0.032142    1.195    2\n",
            "      length[39]    0.022598    0.000014    0.016827    0.028190    0.022319    1.312    2\n",
            "      length[40]    0.014380    0.000007    0.010225    0.018422    0.014286    1.220    2\n",
            "      length[41]    0.019943    0.000013    0.012429    0.024979    0.019857    1.312    2\n",
            "      length[42]    0.012906    0.000013    0.006136    0.018275    0.013949    1.351    2\n",
            "      length[43]    0.017720    0.000029    0.007071    0.026602    0.017329    1.160    2\n",
            "      length[44]    0.011776    0.000013    0.009036    0.017034    0.011149     NA      1 *\n",
            "      length[45]    0.023153    0.000008    0.020225    0.027001    0.023300     NA      1 *\n",
            "      length[46]    0.012280    0.000004    0.010648    0.015147    0.012498     NA      1 *\n",
            "      length[47]    0.016411    0.000005    0.013920    0.018056    0.017257     NA      1 *\n",
            "      length[48]    0.026652    0.000068    0.017286    0.032865    0.029804     NA      1 *\n",
            "      length[49]    0.024940    0.000004    0.022514    0.026319    0.025988     NA      1 *\n",
            "      --------------------------------------------------------------------------------------\n",
            "      + Convergence diagnostic (PSRF = Potential Scale Reduction Factor; Gelman\n",
            "        and Rubin, 1992) should approach 1.0 as runs converge. NA is reported when\n",
            "        deviation of parameter values within all runs is 0 or when a parameter\n",
            "        value (a branch length, for instance) is not sampled in all runs.\n",
            "      * The partition was not found in all runs so the values are unreliable.\n",
            "\n",
            "\n",
            "      Summary statistics for partitions with frequency >= 0.10 in at least one run:\n",
            "          Average standard deviation of split frequencies = 0.059832\n",
            "          Maximum standard deviation of split frequencies = 0.176777\n",
            "          Average PSRF for parameter values (excluding NA and >10.0) = 1.654\n",
            "          Maximum PSRF for parameter values = 1.990\n",
            "\n",
            "\n",
            "      Clade credibility values:\n",
            "\n",
            "      /---------------------------------------------------------- Eudynamys_scolo~ (1)\n",
            "      |                                                                               \n",
            "      |---------------------------------------------------------- Pelecanus_crisp~ (3)\n",
            "      |                                                                               \n",
            "      |                                                   /------ Apis_mellifera_~ (2)\n",
            "      |                                            /--100-+                           \n",
            "      |                                            |      \\------ Drosophila_mel~ (11)\n",
            "      |                               /-----100----+                                  \n",
            "      |                               |            \\------------- Tribolium_casta~ (7)\n",
            "      |                               |                                               \n",
            "      |                               |                   /------ Rattus_norvegic~ (4)\n",
            "      |                         /-100-+      /-----100----+                           \n",
            "      |                         |     |      |            \\------ Mus_musculus_N~ (15)\n",
            "      |                         |     |      |                                        \n",
            "      |                         |     \\--100-+            /------ Myotis_lucifugu~ (6)\n",
            "      |                         |            |     /--100-+                           \n",
            "      |                  /--93--+            |     |      \\------ Pteropus_vampy~ (18)\n",
            "      |                  |      |            \\-100-+                                  \n",
            "      +                  |      |                  \\------------- Ovis_aries_NC_0~ (9)\n",
            "      |                  |      |                                                     \n",
            "      |            /--90-+      |                         /------ Corvus_brachyr~ (14)\n",
            "      |            |     |      \\-----------100-----------+                           \n",
            "      |            |     |                                \\------ Taeniopygia_gu~ (22)\n",
            "      |            |     |                                                            \n",
            "      |            |     \\--------------------------------------- Tyto_alba_MZ31~ (10)\n",
            "      |            |                                                                  \n",
            "      |            |                         /------------------- Gallus_gallus_N~ (5)\n",
            "      |     /--90--+                         |                                        \n",
            "      |     |      |                         |            /------ Anser_cygnoides~ (8)\n",
            "      |     |      |                  /--100-+     /--100-+                           \n",
            "      |     |      |                  |      |     |      \\------ Cygnus_cygnus_~ (21)\n",
            "      |     |      |                  |      \\-100-+                                  \n",
            "      |     |      |                  |            \\------------- Anas_platyrhyn~ (17)\n",
            "      |     |      \\--------93--------+                                               \n",
            "      |     |                         |                   /------ Apteryx_austra~ (12)\n",
            "      \\--88-+                         |            /--100-+                           \n",
            "            |                         |            |      \\------ Dromaius_novae~ (16)\n",
            "            |                         \\-----100----+                                  \n",
            "            |                                      \\------------- Struthio_camel~ (19)\n",
            "            |                                                                         \n",
            "            |                                             /------ Spheniscus_dem~ (13)\n",
            "            |                                      /--100-+                           \n",
            "            |                                      |      \\------ Aptenodytes_fo~ (23)\n",
            "            \\------------------90------------------+                                  \n",
            "                                                   \\------------- Aquila_nipalen~ (20)\n",
            "                                                                                      \n",
            "\n",
            "      Phylogram (based on average branch lengths):\n",
            "\n",
            "      /--- Eudynamys_scolo~ (1)\n",
            "      |                                                                               \n",
            "      |-- Pelecanus_crisp~ (3)\n",
            "      |                                                                               \n",
            "      |                                          /--------------- Apis_mellifera_~ (2)\n",
            "      |                                      /---+                                    \n",
            "      |                                      |   \\--- Drosophila_mel~ (11)\n",
            "      |     /--------------------------------+                                        \n",
            "      |     |                                \\------- Tribolium_casta~ (7)\n",
            "      |     |                                                                         \n",
            "      |     |      /-- Rattus_norvegic~ (4)\n",
            "      |/----+   /--+                                                                  \n",
            "      ||    |   |  \\-- Mus_musculus_N~ (15)\n",
            "      ||    |   |                                                                     \n",
            "      ||    \\---+ /---- Myotis_lucifugu~ (6)\n",
            "      ||        | |                                                                   \n",
            "      ||        | |--- Pteropus_vampy~ (18)\n",
            "      ||        \\-+                                                                   \n",
            "      +|          \\-- Ovis_aries_NC_0~ (9)\n",
            "      ||                                                                              \n",
            "      |+/-- Corvus_brachyr~ (14)\n",
            "      ||+                                                                             \n",
            "      ||\\-- Taeniopygia_gu~ (22)\n",
            "      ||                                                                              \n",
            "      |\\--- Tyto_alba_MZ31~ (10)\n",
            "      |                                                                               \n",
            "      |/--- Gallus_gallus_N~ (5)\n",
            "      ||                                                                              \n",
            "      || /- Anser_cygnoides~ (8)\n",
            "      || |                                                                            \n",
            "      || |- Cygnus_cygnus_~ (21)\n",
            "      ||-+                                                                            \n",
            "      || \\- Anas_platyrhyn~ (17)\n",
            "      |+                                                                              \n",
            "      ||/-- Apteryx_austra~ (12)\n",
            "      |||                                                                             \n",
            "      |||-- Dromaius_novae~ (16)\n",
            "      |\\+                                                                             \n",
            "      | \\-- Struthio_camel~ (19)\n",
            "      |                                                                               \n",
            "      |/- Spheniscus_dem~ (13)\n",
            "      |+                                                                              \n",
            "      |\\- Aptenodytes_fo~ (23)\n",
            "      |                                                                               \n",
            "      \\--- Aquila_nipalen~ (20)\n",
            "                                                                                      \n",
            "      |-------------| 1.000 expected changes per site\n",
            "\n",
            "      Calculating tree probabilities...\n",
            "\n",
            "      Credible sets of trees (4 trees sampled):\n",
            "         90 % credible set contains 2 trees\n",
            "         95 % credible set contains 2 trees\n",
            "         99 % credible set contains 4 trees\n",
            "\n",
            "   Exiting mrbayes block\n",
            "   Reached end of file\n",
            "\n",
            "MrBayes >    End of File encountered on stdin; quitting\n",
            "\n",
            "   Deleting previously defined characters\n",
            "   Deleting previously defined taxa\n",
            "   Quitting program\n",
            "\n",
            "\n",
            "MrBayes error: \n"
          ]
        }
      ],
      "source": [
        "from Bio import Phylo\n",
        "from Bio.Phylo.TreeConstruction import DistanceCalculator, DistanceTreeConstructor\n",
        "import subprocess\n",
        "\n",
        "def construct_tree_NJ(aligned_sequences):\n",
        "    command = [\"megacc\", \"-a\", \"./trees/NJ/mega_nj.mao\", \"-d\", aligned_sequences, \"-o\", \"./trees/NJ/phylo_NJ.tree\"]\n",
        "    subprocess.run(command, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)\n",
        "\n",
        "def construct_tree_ML(aligned_sequences):\n",
        "    command = [\"../../applications/raxml-ng\", \"--search1\", \"--msa\", aligned_sequences, \"--model\", \"GTR+G\", \"--prefix\", \"./trees/ML/phylo_ML\"]\n",
        "    subprocess.run(command)\n",
        "    \n",
        "def construct_tree_BI(genes_tree):\n",
        "    result = subprocess.run(\"mb\", input=f\"execute {genes_tree}\", text=True, capture_output=True)\n",
        "    print(\"MrBayes output:\", result.stdout)\n",
        "    print(\"MrBayes error:\", result.stderr)\n",
        "    \n",
        "construct_tree_NJ(\"./outputs/aligned_sequences.fasta\")\n",
        "construct_tree_ML(\"./outputs/aligned_sequences.fasta\")\n",
        "construct_tree_BI(\"./trees/BI/mb_bi.nex\")"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Y-7QslLTk1s7"
      },
      "source": [
        "### Step 6: In-Depth Phylogenetic Tree Visualization\n",
        "\n",
        "Having constructed phylogenetic trees using different methods, our next task is to visualize these trees effectively. This step is crucial for interpreting the results and communicating our findings.\n",
        "\n",
        "#### Visualization Tools:\n",
        "\n",
        "1. **FigTree**:\n",
        "    - **Overview**: FigTree is designed for the graphical representation of phylogenetic trees. It's excellent for creating publication-ready visualizations.\n",
        "    - **Resource**: [FigTree Tool](http://tree.bio.ed.ac.uk/software/figtree/)\n",
        "    - **Usage**: Use FigTree to add detailed annotations, adjust branch colors, and format tree layouts for clear, interpretable visualizations.\n",
        "\n",
        "2. **iTOL (Interactive Tree Of Life)**:\n",
        "    - **Overview**: iTOL is a web-based tool for the display, annotation, and management of phylogenetic trees, offering extensive customization options.\n",
        "    - **Resource**: [iTOL Website](https://itol.embl.de/)\n",
        "    - **Usage**: Ideal for interactive tree visualizations. It allows users to explore different layers of data through their tree, such as adding charts or color-coding branches.\n",
        "\n",
        "3. **Dendroscope**:\n",
        "    - **Overview**: Dendroscope is a software program for viewing and editing phylogenetic trees, particularly useful for large datasets.\n",
        "    - **Resource**: [Dendroscope Download](https://uni-tuebingen.de/fakultaeten/mathematisch-naturwissenschaftliche-fakultaet/fachbereiche/informatik/lehrstuehle/algorithms-in-bioinformatics/software/dendroscope/)\n",
        "    - **Usage**: Utilize Dendroscope when dealing with large and complex trees or when you need to compare multiple trees side-by-side.\n",
        "\n",
        "#### Task:\n",
        "\n",
        "- **Visualize Each Tree**: Use one or more of the above tools to visualize the phylogenetic trees you constructed using Bayesian inference, maximum likelihood, and neighbor-joining methods.\n",
        "- **Highlight Differences**: Focus on highlighting the differences and similarities between the trees obtained from the different methods. Pay attention to tree topology, branch lengths, and any notable patterns.\n",
        "- **Interpretation and Presentation**: Aim for visualizations that are not only accurate but also interpretable and visually appealing. This will enhance the clarity of your work."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "LBXt-2Q3k1s7"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "javax.swing.UIManager$LookAndFeelInfo[Metal javax.swing.plaf.metal.MetalLookAndFeel]\n",
            "javax.swing.UIManager$LookAndFeelInfo[Nimbus javax.swing.plaf.nimbus.NimbusLookAndFeel]\n",
            "javax.swing.UIManager$LookAndFeelInfo[CDE/Motif com.sun.java.swing.plaf.motif.MotifLookAndFeel]\n",
            "javax.swing.UIManager$LookAndFeelInfo[Mac OS X com.apple.laf.AquaLookAndFeel]\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "CompletedProcess(args=['java', '-jar', '../../applications/figtree.jar'], returncode=0)"
            ]
          },
          "execution_count": 3,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "import subprocess\n",
        "\n",
        "\n",
        "itol_urls = {\"NJ_tree\": \"https://itol.embl.de/tree/91235234219193121707301647\", \n",
        "             \"ML_tree\": \"https://itol.embl.de/tree/91235234219159851707317778\",\n",
        "             \"BI_tree\": \"https://itol.embl.de/tree/91235234219157691707317764\"}\n",
        "\n",
        "subprocess.run([\"java\", \"-jar\", \"../../applications/figtree.jar\"])"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "P6Ta6GCwk1s7"
      },
      "source": [
        "### Cross-Disciplinary Applications (Optional)\n",
        "\n",
        "This is an optional part with bonus, relative to the depth of your analysis. Refer to the first part of this notebook. You have complete freedom to do this part anyway you like, but to gain a portion of the bonus score for this section, a bare minimum effort is required."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "nV5UFbZmk1s8"
      },
      "source": [
        "### Conclusion and Reflective Insights\n",
        "\n",
        "As we conclude our exploration of phylogenetic tree construction and analysis, let's reflect on the insights learned from this task and consider questions that emerge from our findings.\n",
        "\n",
        "#### Interpretation of Results:\n",
        "\n",
        "- Reflect on the phylogenetic trees produced by each method (Bayesian inference, maximum likelihood, and neighbor-joining). Consider how the differences in tree topology might offer varied perspectives on the evolutionary relationships among the species.\n",
        "\n",
        "#### Questions to Ponder:\n",
        "\n",
        "1. **Species Divergence**: Based on the trees, which species appear to have the most ancient divergence? How might this information contribute to our understanding of their evolutionary history?\n",
        "   \n",
        "2. **Common Ancestors**: Are there any unexpected pairings or groupings of species that suggest a closer evolutionary relationship than previously thought? How could this reshape our understanding of these species' evolutionary paths?\n",
        "\n",
        "3. **Methodology Insights**: Considering the discrepancies between the trees generated by different methods, what might this tell us about the limitations and strengths of each phylogenetic analysis method?\n",
        "\n",
        "4. **Conservation Implications**: Considering the evolutionary relationships revealed in your phylogenetic analysis, what insights can be gained for conservation strategies? Specifically, how could understanding the close evolutionary ties between species, which might be facing distinct environmental challenges, guide targeted conservation efforts?\n",
        "\n",
        "<blockquote style=\"font-family:Arial; color:red; font-size:16px; border-left:0px solid red; padding: 10px;\">\n",
        "    <strong>Don't forget to answer these questions!</strong>\n",
        "</blockquote>\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.3"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
